Feature/road mesh

Sidlertime opened the pull request:
* zur Optimierung der Performance ein Mesh mit entsprechenden Vertices zu benutzen anstelle von vielen einzelnen Cubes, die zurechtgescaled sind
* Material mit Textur für die Straße

Sidlertime committed:
make one big mesh for the road instead of million cubes and apply texture

Sidlertime committed:
commenting

Sidlertime committed:
using ROAD_HEIGHT, changing texture and tiling to fit

funketh commented on Assets/Scripts/DataTypes/Edges/Edge.cs:
```diff
@@ -43,20 +43,56 @@ private Edge(Edge other, List<Lane> outgoingLanes)
+            };
+
+            // spawn the road and apply Mesh and Material with adapted tiling
+            var road = Object.Instantiate(prefab, Vector3.zero, Quaternion.identity);
```
bitte gameObject nutzen (s. #66)

funketh replied:
https://github.com/rrickfox/traffic-sim/pull/65/commits/5059c170e89cfba88db47df1db2eb33aa558fd54

Sidlertime committed:
using edge.gameObject instead of a new one

Sidlertime committed:
creating a second road and fixing material-tiling to be individual

funketh committed:
shorten mesh construction

funketh commented on Assets/Scripts/DataTypes/Edges/Edge.cs:
```diff
@@ -43,20 +43,48 @@ private Edge(Edge other, List<Lane> outgoingLanes)
+            gameObject.GetComponent<MeshFilter>().mesh = new Mesh {vertices=vertices, triangles=triangles, uv=uvs};
+            var tiling = shape.length * CONSTANTS.DISTANCE_UNIT / 12f;
+            gameObject.GetComponent<MeshRenderer>().material.SetTextureScale("_MainTex", new Vector2(1, tiling));
         }
```
vielleicht die vertices umbenennen, verwirrt etwas, da nicht die Vertex klasse gemeint ist

RichardKarlFuchs replied:
Es handelt sich aber um die korrekte Bezeichnung im Kontext von Meshes. Vielleicht MeshVertex / MeshVertices?

funketh replied:
ja das wär besser denk ich

RichardKarlFuchs replied:
b7512ff

RichardKarlFuchs committed:
Merge branch 'develop' into feature/road_mesh

RichardKarlFuchs committed:
remove roadPrefab from EndPoint

Sidlertime committed:
change names "vertices" into "meshVertices" and "vertexIndex" into "meshVertexIndex"

Sidlertime committed:
make tiling as integer to end the roadtexture with end of a tile

funketh committed:
generate mesh vertices using lists

- no need to keep track of indices that way
- triangles is generated independently using LINQ

funketh commented on Assets/Scripts/PointPlacer.cs:
```diff
@@ -19,12 +21,52 @@ void Start()
         {
-            var g = GameObject.CreatePrimitive(PrimitiveType.Cube);
-            g.transform.position = new Vector3(p.position.x, 0, p.position.y);
-            g.transform.rotation = Quaternion.Euler(0, Vector2.SignedAngle(p.forward, Vector2.right), 0);
-            g.transform.localScale = new Vector3(CONSTANTS.DISTANCE_UNIT, 0.1f, CONSTANTS.LANE_WIDTH);
+            var p = shape.points[i];
+            var left = new Vector2(-p.forward.y, p.forward.x);
+            var newPosLeft = p.position + left * CONSTANTS.LANE_WIDTH;
+            var newPosRight = p.position - left * CONSTANTS.LANE_WIDTH;
+            vertices[vertexIndex] = new Vector3(newPosLeft.x, 0.025f, newPosLeft.y);
+            vertices[vertexIndex + 1] = new Vector3(newPosRight.x, 0.025f, newPosRight.y);
+
+            var relativPos = i / (float)(shape.points.Length-1);
+            uvs[vertexIndex] = new Vector2(0f, relativPos);
+            uvs[vertexIndex + 1] = new Vector2(1f, relativPos);
+
+            if (i < shape.points.Length-1)
+            {
+                triangles[triangleIndex] = vertexIndex;
+                triangles[triangleIndex + 1] = vertexIndex + 2;
+                triangles[triangleIndex + 2] = vertexIndex + 1;
+
+                triangles[triangleIndex + 3] = vertexIndex + 1;
+                triangles[triangleIndex + 4] = vertexIndex + 2;
+                triangles[triangleIndex + 5] = vertexIndex + 3;
+            }
+
+            vertexIndex += 2;
+            triangleIndex += 6;
         }
+
+        var mesh = new Mesh
+        {
+            vertices = vertices,
+            triangles = triangles,
+            uv = uvs
+        };
+
+        var g = GameObject.CreatePrimitive(PrimitiveType.Cube);
+        g.GetComponent<MeshFilter>().mesh = mesh;
```
was genau wird hier gemacht?
Sieht sehr redundant zu `Edge.Display()` aus

RichardKarlFuchs replied:
Handelt sich um eine Testszene mit Testscript, ist für den eigentlichen Code irrelevant. Wird mit feature/road_texture entfernt.
